prov/efa/src/rxr/rxr_read.c

```diff
@@ -235,18 +294,67 @@ int rxr_read_post_or_queue(struct rxr_ep *ep, int entry_type, void *x_entry)
 		return -FI_ENOBUFS;
 	}
 
-	err = rxr_read_post(ep, read_entry);
-	if (err == -FI_EAGAIN) {
-		dlist_insert_tail(&read_entry->pending_entry, &ep->read_pending_list);
-		read_entry->state = RXR_RDMA_ENTRY_PENDING;
-		err = 0;
-	} else if(err) {
-		rxr_read_release_entry(ep, read_entry);
+	return rxr_read_post_or_queue(ep, read_entry);
+}
+
+int rxr_read_post_local_read_or_queue(struct rxr_ep *ep,
+				      struct rxr_rx_entry *rx_entry,
+				      size_t data_offset,
+				      struct rxr_pkt_entry *pkt_entry,
+				      char *data, size_t data_size)
+{
+	int err;
+	struct rxr_read_entry *read_entry;
+
+	read_entry = ofi_buf_alloc(ep->read_entry_pool);
+	if (!read_entry) {
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "RDMA entries exhausted\n");
+		return -FI_ENOBUFS;
+	}
+
+	assert(ep->rdm_self_addr != FI_ADDR_NOTAVAIL);
```

Comment:
This is duplicating code that's already in the rx/tx entry init and handling logic. It would be much cleaner if we can avoid the `rxr_read_entry` structure but it seems like you're using it to create another iov to consume, right? We should rethink this later.

Reply:
Yes, I think eventually we want to merge tx_entry, rx_entry and read_entry into one structure.